gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/watermarker/PartitionLevelWatermarker.java

```diff
@@ -0,0 +1,339 @@
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+package gobblin.data.management.conversion.hive.watermarker;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.annotation.Nonnull;
+
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.hadoop.hive.ql.metadata.Partition;
+import org.apache.hadoop.hive.ql.metadata.Table;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Function;
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Collections2;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Ordering;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.SourceState;
+import gobblin.configuration.WorkUnitState;
+import gobblin.data.management.conversion.hive.converter.AbstractAvroToOrcConverter;
+import gobblin.source.extractor.Watermark;
+import gobblin.source.extractor.WatermarkInterval;
+import gobblin.source.extractor.extract.LongWatermark;
+import gobblin.source.workunit.WorkUnit;
+
+
+/**
+ * A {@link HiveSourceWatermarker} that maintains watermarks for each {@link Partition}. For every {@link Table} it creates
+ *  a NoOp {@link WorkUnit} that stores watermarks for all its {@link Partition}s. The Noop workunit is identified by the
+ *  property {@link #IS_WATERMARK_WORKUNIT_KEY} being set to true.
+ *  <p>
+ *  The watermark is stored as a {@link MultiKeyValueLongWatermark} which is an extension to {@link Map} that implements
+ *  gobblin's {@link Watermark}. The key is a {@link Partition} identifier and value is the watermark for this {@link Partition}
+ *  </p>
+ *  <p>
+ *  The maximum number of {@link Partition} watermarks for each {@link Table} can be adjusted by the property
+ *  {@link #MAX_PARTITIONS_IN_WATERMARK_PER_DATASET_KEY},
+ *  The default is {@value #DEFAULT_MAX_PARTITIONS_IN_WATERMARK_PER_DATASET}. If a previous watermark is not found for
+ *  as partition, it returns 0 as the watermark
+ *  </p>
+ */
+@Slf4j
+public class PartitionLevelWatermarker implements HiveSourceWatermarker {
+
+  public static final String IS_WATERMARK_WORKUNIT_KEY = "hive.source.watermark.isWatermarkWorkUnit";
+  public static final String MAX_PARTITIONS_IN_WATERMARK_PER_DATASET_KEY = "hive.source.watermark.maxPartitions";
+  public static final int DEFAULT_MAX_PARTITIONS_IN_WATERMARK_PER_DATASET = 50;
+
+  private static final Joiner PARTITION_VALUES_JOINER = Joiner.on(",");
+
+  /**
+   * Ordering used to retain only the {@link #maxPartitionsPerDataset} allowed
+   */
+  private static final Ordering<Map.Entry<String, Long>> MOST_RECENTLY_UPDATED_PARTITIONS_ORDERING = Ordering.natural()
```

Comment:
In line `247` we pick the `greatestOf` instead of `leastOf` `MOST_RECENTLY_UPDATED_PARTITIONS_ORDERING.greatestOf(partitionWatermarks.entrySet(), this.maxPartitionsPerDataset)`